content/guides/java/inject-extract.md

```diff
@@ -2,8 +2,142 @@
+        String operationName) {
+    // format the headers for extraction
+    MultivaluedMap<String, String> rawHeaders = httpHeaders.getRequestHeaders();
+    final HashMap<String, String> headers = new HashMap<String, String>();
+    for (String key : rawHeaders.keySet()) {
+        headers.put(key, rawHeaders.get(key).get(0));
+    }
+
+    Tracer.SpanBuilder spanBuilder;
+    try {
+        SpanContext parentSpan = tracer.extract(Format.Builtin.HTTP_HEADERS, new TextMapExtractAdapter(headers));
+        if (parentSpan == null) {
+            spanBuilder = tracer.buildSpan(operationName);
+        } else {
+            spanBuilder = tracer.buildSpan(operationName).asChildOf(parentSpan);
+        }
+    } catch (IllegalArgumentException e) {
+        spanBuilder = tracer.buildSpan(operationName);
+    }
+    return spanBuilder.withTag(Tags.SPAN_KIND.getKey(), Tags.SPAN_KIND_SERVER).startActive(true);
+}
+```
+
+In the above example, instead of using a dedicated adapter class to convert JAXRS `HttpHeaders` type into `io.opentracing.propagation.TextMap`, the headers are copied to a plain `HashMap<String, String>` and converted using a standard adapter `TextMapExtractAdapter`.
+
+## Injecting and Extracting TextMap
+The process of injecting and extracting `TextMap` is similar to that of HTTP. Given below are some elaborated code examples of injecting and extracting tracing information using `TextMap` format.
+
+```
+protected void attachTraceInfo(Tracer tracer, Span span, final Invocation inv) {
+    tracer.inject(span.context(), Format.Builtin.TEXT_MAP, new TextMap() {
+
+        @Override
+        public void put(String key, String value) {
+            inv.getAttachments().put(key, value);
+        }
+
+        @Override
+        public Iterator<Map.Entry<String, String>> iterator() {
+            throw new UnsupportedOperationException("TextMapInjectAdapter should only be used with Tracer.inject()");
+        }
+    });
+}
+```
+The above method injects the invocation information into the span’s context on the consumer side. The `TextMap` interface has been implemented to create a `TextMap` of invocation attachments.
+
+```
+protected Span extractTraceInfo(Tracer tracer, Invoker<?> invoker, Invocation inv) {
+
+    Tracer.SpanBuilder span = tracer.buildSpan(“Operation_Name_Here”);
+    try {
+        SpanContext spanContext = tracer.extract(Format.Builtin.TEXT_MAP, new TextMapExtractAdapter(inv.getAttachments()));
+        if (spanContext != null) {
+            //if spanContext is extracted, the spanContext is propagated to the new span
+              
+            span.asChildOf(spanContext);
+        }
+    } catch (Exception e) {
+        span.withTag("Error", "extract from request fail, error msg:" + e.getMessage());
+    }
+    return span.startManual();
+}
+```
+On the provider side, `tracer.extract()` is used to extract the invocation attachments. The logic is similar to `tracer.inject()`.
```

Comment:
s/invocation attachments/`SpanContext`/

s/is similar to `tracer.inject()`/is the reverse operation of the `tracer.inject()`/